### PR TITLE
Quadrature training strategy

### PR DESCRIPTION
**Pull request by KirillZubov**

### CODE REVIEW

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -450,11 +466,13 @@ function DiffEqBase.discretize(pde_system::PDESystem, discretization::PhysicsInf
 
     pde_loss_function = get_loss_function(eval(expr_pde_loss_function),
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:467]: for safety, you probably want to use RuntimeGeneratedFunctions.jl here for the eval.

**Comment by KirillZubov** [comment]
`@RuntimeGeneratedFunction` doesn't work with inner functions.

```julia
using RuntimeGeneratedFunctions
RuntimeGeneratedFunctions.init(@__MODULE__)

expr_func =
:((x)->begin
      u = (x)->begin x+1 end
      u(1)+1
end)

f = @RuntimeGeneratedFunction(expr_func)
f(1)

ERROR: The function body AST defined by this @generated function is not pure. This likely means it contains a closure or comprehension.
Stacktrace:
 [1] (::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{var"#_RuntimeGeneratedFunctions_ModTag",(0xb9, 0x91, 0x34, 0xab, 0x78, 0x06, 0xb6, 0x9e, 0xbc, 0xe9, 0xc0, 0xe0, 0x85, 0xdd, 0xdb, 0xdd, 0x14, 0x0e, 0xae, 0xd2, 0xba, 0xc9, 0x3f, 0x30, 0x33, 0xba, 0x01, 0xaa, 0x1f, 0x4b, 0x00, 0x60, 0xfa, 0x19, 0xcd, 0x0f, 0xf1, 0x3b, 0xff, 0x19, 0x8d, 0x71, 0x79, 0x01, 0x30, 0x21, 0x43, 0x77, 0x89, 0x98, 0xb5, 0xa7, 0x62, 0x8e, 0x77, 0x05, 0x61, 0xde, 0xb3, 0x77, 0x4c, 0x4e, 0x7e, 0x98),(:x,)})(::Int64) at /Users/kirill/.julia/packages/RuntimeGeneratedFunctions/fIcZp/src/RuntimeGeneratedFunctions.jl:68
 [2] top-level scope at none:1
```

**Comment by ChrisRackauckas** [comment]
How are the inner functions showing up?

**Comment by KirillZubov** [comment]
@ChrisRackauckas when it's used `build_loss_function`. 

```julia
expr_pde_loss_function = build_loss_function(eqs,indvars,depvars,
                                                 dict_indvars,dict_depvars)
julia> expr_pde_loss_function
:((phi, cord, θ, derivative)->begin
          #= none:35 =#
          #= none:35 =#
          begin
              begin
                  u = ((x, y, θ)->begin
                              #= none:17 =#
                              (phi([x, y], θ))[1]
                          end)
                  u_d = ((cord, θ)->begin
                              #= none:21 =#
                              (phi(cord, θ))[1]
                          end)
              end
              let (x, y) = (cord[1], cord[2])
                  [(derivative(u_d, [x, y], Array{Float32,1}[[0.0049215667, 0.0], [0.0049215667, 0.0]], 2, θ) 
                   + derivative(u_d, [x, y], Array{Float32,1}[[0.0, 0.0049215667], [0.0, 0.0049215667]], 2, θ)) - -(sin(π * x)) * sin(π * y)]
              end
          end
      end)
```

**Comment by ChrisRackauckas** [comment]
Can't `u` and `u_d` be passed in and enclosed instead?

**Comment by KirillZubov** [comment]
> Can't `u` and `u_d` be passed in and enclosed instead?

yes, it would be something like this:
```julia
u_arr = []
u_d_arr = []
for i in eachindex(depvars)
    u = @RuntimeGeneratedFunction(:((x, y, θ; phi=phi,derivative=derivative)->begin
        (phi([x, y], θ))[1]
    end))
    push!(u_arr,u)
    u_d = @RuntimeGeneratedFunction(:((cord, θ; phi=phi, derivative=derivative)->begin
        (phi(cord, θ))[1]
    end))
    push!(u_d_arr,u_d)
end

loss_functionss = @RuntimeGeneratedFunction(:((phi, cord, θ, derivative, u_arr, u_d_arr)->begin
    (u,) = (u_arr[1]),)
    (u_d,) = (u_d_arr[1]),)
    let (x, y) = (cord[1], cord[2])
        [derivative(u_d, [x, y], Array{Float32,1}[[0.0049215667, 0.0], [0.0049215667, 0.0]], 2, θ) + u(x, y, θ) - (sin(π * x)) * sin(π * y)]
    end
end))
```

**Comment by KirillZubov** [comment]
it was required to move the discretization `dx` from PhysicsInformed to directly `TrainingStrategies`.  Parameter of discretization doesn't need for `QuadratuteTraning` but need for `GridTraning` and `StochasticTraning`.

**Comment by ChrisRackauckas** [comment]
It's not needed for StochasticTraining either?

**Comment by KirillZubov** [comment]
> It's not needed for StochasticTraining either?

it is needed in StochasticTraining for generating a training set.

**Comment by ChrisRackauckas** [comment]
Why not just uniformly sample from the domain?

**Comment by KirillZubov** [comment]
Descritization determines how many points are in the training set and affects the prediction accuracy and execution time. We can write a small algorithm that set optimal descritization if it's not specified in interact: `StochasticTraining(nothing)`, but I would not completely remove dx from the StochasticTraining interface. and for GridTraning it's the same.

**Comment by ChrisRackauckas** [comment]
GridTraining should take a `dx`. `StochasticTraining` should take a number of points to sample. `QuadratureTraining` should take the arguments for the quadarture.

**Pull request merged**